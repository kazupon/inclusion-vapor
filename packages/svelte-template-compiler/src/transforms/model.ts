// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vModel.ts

import { BindingTypes, createSimpleExpression, isMemberExpression } from '@vue-vapor/compiler-dom'
import { IRNodeTypes } from '../ir/index.ts'
import { getExpSource } from './utils.ts'

import type { VaporHelper } from '@vue-vapor/compiler-vapor'
import type { DirectiveTransform } from './types.ts'

// TODO: transform vapor v-model from svelte some bindings
// https://svelte.dev/docs/element-directives#bind-property
// https://svelte.dev/docs/element-directives#binding-select-value
// ... and more on https://svelte.dev/docs/element-directives
export const transformVModel: DirectiveTransform = (dir, node, context) => {
  console.log('transformVModel', dir, node)

  const { exp, arg } = dir
  if (!exp) {
    // TODO: should throw error on `onError`
    return
  }

  // we assume v-model directives are always parsed
  // (not artificially created by a transform)
  const rawExp = exp.loc.source
  console.log('rawExp', rawExp)
  console.log('bindingMetadata', context.options.bindingMetadata)

  // TODO: do we need to handle from `bindingMetadata`?
  const bindingType = (context.options.bindingMetadata || {})[rawExp]
  console.log('bindingType', bindingType)

  // check props
  if (bindingType === BindingTypes.PROPS || bindingType === BindingTypes.PROPS_ALIASED) {
    // TODO: should throw error on `onError`
    return
  }

  const expString = exp.content
  const maybeRef =
    !__BROWSER__ &&
    context.options.inline &&
    (bindingType === BindingTypes.SETUP_LET ||
      bindingType === BindingTypes.SETUP_REF ||
      bindingType === BindingTypes.SETUP_MAYBE_REF)
  console.log('maybeRef', maybeRef)
  console.log('expString', expString)

  if (!expString.trim() || (!isMemberExpression(getExpSource(exp), context.options) && !maybeRef)) {
    // TODO: should throw error on `onError`
    return
  }

  const isComponent = node.type === 'InlineComponent'
  if (isComponent) {
    return {
      // eslint-disable-next-line unicorn/prefer-logical-operator-over-ternary
      key: arg ? arg : createSimpleExpression('modelValue', true),
      value: exp,
      model: true,
      // @ts-expect-error -- FIXME should resolve `modifiers` type
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      modelModifiers: dir.modifiers.map(m => m.content)
    }
  }

  const { name: tag } = node
  const isCustomElement = context.options.isCustomElement(tag)

  let runtimeDirective: VaporHelper | undefined = 'vModelText'
  if (tag === 'input' || tag === 'textarea' || tag === 'select' || isCustomElement) {
    if (tag === 'input' || isCustomElement) {
      /*
      const type = findProp(node, 'type')
      if (type) {
        if (type.type === NodeTypes.DIRECTIVE) {
          // :type="foo"
          runtimeDirective = 'vModelDynamic'
        } else if (type.value) {
          switch (type.value.content) {
            case 'radio':
              runtimeDirective = 'vModelRadio'
              break
            case 'checkbox':
              runtimeDirective = 'vModelCheckbox'
              break
            case 'file':
              runtimeDirective = undefined
              context.options.onError(
                createDOMCompilerError(
                  DOMErrorCodes.X_V_MODEL_ON_FILE_INPUT_ELEMENT,
                  dir.loc,
                ),
              )
              break
            default:
              // text type
              __DEV__ && checkDuplicatedValue()
              break
          }
        }
      } else if (hasDynamicKeyVBind(node)) {
        // element has bindings with dynamic keys, which can possibly contain
        // "type".
        runtimeDirective = 'vModelDynamic'
      } else {
        // text type
        __DEV__ && checkDuplicatedValue()
      }
        */
    } else if (tag === 'select') {
      runtimeDirective = 'vModelSelect'
    } else {
      // textarea
      // __DEV__ && checkDuplicatedValue()
    }
  } else {
    // TODO:
    // context.options.onError(
    //   createDOMCompilerError(
    //     DOMErrorCodes.X_V_MODEL_ON_INVALID_ELEMENT,
    //     dir.loc,
    //   ),
    // )
  }

  context.registerOperation({
    type: IRNodeTypes.SET_MODEL_VALUE,
    element: context.reference(),
    key: arg || createSimpleExpression('modelValue', true),
    value: exp,
    isComponent
  })

  if (runtimeDirective) {
    context.registerOperation({
      type: IRNodeTypes.WITH_DIRECTIVE,
      element: context.reference(),
      dir,
      name: runtimeDirective,
      builtin: true
    })
  }

  // NOTE: we don't need duplicate value check, because svelte binding allows it
  // function checkDuplicatedValue() {
  //   const value = findDir(node, 'bind')
  //   if (value && isStaticArgOf(value.arg, 'value')) {
  //     context.options.onError(
  //       createDOMCompilerError(
  //         DOMErrorCodes.X_V_MODEL_UNNECESSARY_VALUE,
  //         value.loc,
  //       ),
  //     )
  //   }
  // }
}
