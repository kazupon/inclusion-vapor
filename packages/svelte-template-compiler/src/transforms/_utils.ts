import { parse } from 'svelte/compiler'
import { generate } from '@vue-vapor/compiler-vapor'
import { transform } from '../transform'
import { IRNodeTypes } from '../ir'

import type {
  RootIRNode as VaporRootIRNode,
  CompilerOptions as VaporCompilerOptions
} from '@vue-vapor/compiler-vapor'
import type { CompilerOptions } from '../compile'
import type { RootNode, RootIRNode } from '../ir'

export const DEFAULT_OPTIONS: CompilerOptions = {
  prefixIdentifiers: true
}
export const DEFAULT_VAPOR_COMPILER_OPTIONS = DEFAULT_OPTIONS as VaporCompilerOptions

export function makeCompile(options: CompilerOptions = {}) {
  return (
    source: string,
    overrideOptions: CompilerOptions = {}
  ): {
    ast: RootNode
    ir: RootIRNode
    code: string
    helpers: Set<string>
    vaporHelpers: Set<string>
  } => {
    const svelteAst = parse(source)

    const ast: RootNode = {
      type: IRNodeTypes.ROOT,
      children: svelteAst.html.children || [],
      source,
      components: [],
      directives: [],
      helpers: new Set(),
      temps: 0
    }

    const ir = transform(ast, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })
    const { code, helpers, vaporHelpers } = generate(ir as unknown as VaporRootIRNode, {
      ...DEFAULT_OPTIONS,
      ...options,
      ...overrideOptions
    })

    return { ast, ir, code, helpers, vaporHelpers }
  }
}
