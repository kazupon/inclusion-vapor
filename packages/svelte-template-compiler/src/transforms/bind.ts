// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vBind.ts

import {
  ErrorCodes,
  NodeTypes,
  createCompilerError,
  createSimpleExpression
} from '@vue-vapor/compiler-dom'
import { camelize, extend } from '@vue-vapor/shared'
import { isReservedProp, resolveExpression } from './utils.ts'

import type { SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type { TransformContext } from './context.ts'
import type { DirectiveTransform } from './types.ts'

// same-name shorthand - :arg is expanded to :arg="arg"
export function normalizeBindShorthand(
  arg: SimpleExpressionNode,
  context: TransformContext
): SimpleExpressionNode {
  if (arg.type !== NodeTypes.SIMPLE_EXPRESSION || !arg.isStatic) {
    // only simple expression is allowed for same-name shorthand
    context.options.onError(
      createCompilerError(ErrorCodes.X_V_BIND_INVALID_SAME_NAME_ARGUMENT, arg.loc)
    )
    return createSimpleExpression('', true, arg.loc)
  }

  const propName = camelize(arg.content)
  const exp = createSimpleExpression(propName, false, arg.loc)
  exp.ast = null // eslint-disable-line unicorn/no-null
  return exp
}

export const transformVBind: DirectiveTransform = (dir, _node, context) => {
  const { loc, modifiers } = dir
  let { exp } = dir
  let arg = dir.arg!
  const modifiersString = new Set(modifiers.map(s => s.content))

  if (!exp) {
    exp = normalizeBindShorthand(arg, context)
  }
  if (!exp.content.trim()) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
    exp = createSimpleExpression('', true, loc)
  }

  exp = resolveExpression(exp)
  arg = resolveExpression(arg)

  if (arg.isStatic && isReservedProp(arg.content)) {
    return
  }

  let camel = false
  if (modifiersString.has('camel')) {
    if (arg.isStatic) {
      arg = extend({}, arg, { content: camelize(arg.content) })
    } else {
      camel = true
    }
  }

  return {
    key: arg,
    value: exp,
    loc,
    runtimeCamelize: camel,
    modifier: modifiersString.has('prop') ? '.' : modifiersString.has('attr') ? '^' : undefined
  }
}
