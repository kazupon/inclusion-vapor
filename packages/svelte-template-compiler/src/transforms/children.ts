// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformChildren.ts

import { DynamicFlag, IRNodeTypes, isSvelteText } from '../ir/index.ts'
import { transformNode } from '../transform.ts'

import type { IRDynamicInfo, SvelteElement } from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { NodeTransform } from './types.ts'

export const transformChildren: NodeTransform = (node, context) => {
  const isFragment =
    node.type === IRNodeTypes.ROOT ||
    node.type === 'Fragment' ||
    node.type === 'InlineComponent' ||
    node.type === 'Slot'
  if (node.type !== 'Element' && !isFragment) {
    return
  }
  processChildren(node, context, isFragment)
}

export function processChildren(
  node: Parameters<NodeTransform>[0],
  context: Parameters<NodeTransform>[1],
  isFragment: boolean
): ReturnType<NodeTransform> {
  const children = node.children || []

  // normalize svelte text
  for (const [index, child] of [...children].entries()) {
    if (isSvelteText(child) && !child.data.trim()) {
      child.data = ' '
      if (!index) {
        children.splice(0, 1)
      } else if (index === children.length) {
        children.splice(-1, 1)
      }
    }
  }

  for (const [i, child] of children.entries()) {
    const childContext = context.create(child, i)
    transformNode(childContext)

    if (isFragment) {
      childContext.reference()
      childContext.registerTemplate()

      if (
        !(childContext.dynamic.flags & DynamicFlag.NON_TEMPLATE) ||
        childContext.dynamic.flags & DynamicFlag.INSERT
      ) {
        context.block.returns.push(childContext.dynamic.id!)
      }
    } else {
      context.childrenTemplate.push(childContext.template)
    }

    context.dynamic.children[i] = childContext.dynamic
  }

  if (!isFragment) {
    processDynamicChildren(context as TransformContext<SvelteElement>)
  }
}

function processDynamicChildren(context: TransformContext<SvelteElement>) {
  let prevDynamics: IRDynamicInfo[] = []
  let hasStaticTemplate = false
  const children = context.dynamic.children

  for (const [index, child] of children.entries()) {
    if (child.flags & DynamicFlag.INSERT) {
      prevDynamics.push(child)
    }

    if (!(child.flags & DynamicFlag.NON_TEMPLATE)) {
      if (prevDynamics.length > 0) {
        if (hasStaticTemplate) {
          context.childrenTemplate[index - prevDynamics.length] = `<!>`

          prevDynamics[0].flags -= DynamicFlag.NON_TEMPLATE
          const anchor = (prevDynamics[0].anchor = context.increaseId())

          context.registerOperation({
            type: IRNodeTypes.INSERT_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference(),
            anchor
          })
        } else {
          context.registerOperation({
            type: IRNodeTypes.PREPEND_NODE,
            elements: prevDynamics.map(child => child.id!),
            parent: context.reference()
          })
        }
        prevDynamics = []
      }
      hasStaticTemplate = true
    }
  }

  if (prevDynamics.length > 0) {
    context.registerOperation({
      type: IRNodeTypes.INSERT_NODE,
      elements: prevDynamics.map(child => child.id!),
      parent: context.reference()
    })
  }
}
