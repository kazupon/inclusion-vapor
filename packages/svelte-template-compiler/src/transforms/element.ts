// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/transformElement.ts

import {
  createCompilerError,
  createSimpleExpression,
  ErrorCodes,
  isStaticArgOf,
  NodeTypes
} from '@vue-vapor/compiler-dom'
import { camelize, capitalize, extend, isVoidTag } from '@vue-vapor/shared'
import { isValidHTMLNesting } from '../htmlNesting.ts'
import {
  convertProps,
  convertVaporDirectiveComponentExpression,
  DynamicFlag,
  IRDynamicPropsKind,
  IRNodeTypes,
  isSvelteComment,
  isSvelteComponentTag,
  isSvelteElement
} from '../ir/index.ts'
import { EMPTY_EXPRESSION, isReservedProp } from './utils.ts'

import type { AttributeNode, SimpleExpressionNode } from '@vue-vapor/compiler-dom'
import type {
  IRProp,
  IRProps,
  IRPropsDynamicAttribute,
  IRPropsStatic,
  SvelteComponentTag,
  SvelteElement,
  VaporDirectiveNode
} from '../ir/index.ts'
import type { TransformContext } from './context.ts'
import type { DirectiveTransformResult } from './index.ts'
import type { NodeTransform } from './types.ts'

export const transformElement: NodeTransform = (_node, context) => {
  if (__DEV__) {
    console.log(
      'transformElement',
      _node.type,
      // @ts-expect-error -- IGNORE
      _node?.name
    )
  }

  return function postTransformElement() {
    const { node } = context
    if (!isSvelteElement(node)) {
      return
    }
    if (node.type === 'Slot' || node.type === 'SlotTemplate') {
      return
    }

    const isComponent = node.type === 'InlineComponent'
    const isDynamicComponent = isSvelteComponentTag(node)

    // apply scoped style
    context.applyScopedCss(node)

    const propsResult = buildProps(
      node,
      context as TransformContext<SvelteElement>,
      isComponent,
      isDynamicComponent
    )

    // TODO:
    /*
    let { parent } = context
    while (
      parent &&
      parent.parent &&
      parent.node.type === NodeTypes.ELEMENT &&
      parent.node.tagType === ElementTypes.TEMPLATE
    ) {
      parent = parent.parent
    }
    const singleRoot =
      context.root === parent &&
      parent.node.children.filter(child => child.type !== NodeTypes.COMMENT)
        .length === 1
    */
    const { parent } = context
    const singleRoot =
      context.root === parent &&
      (parent.node.children || []).filter(child => !isSvelteComment(child)).length === 1

    ;(isComponent ? transformComponentElement : transformNativeElement)(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-argument
      node as any,
      propsResult,
      singleRoot,
      context as TransformContext<SvelteElement>,
      isDynamicComponent
    )
  }
}

function transformComponentElement(
  node: SvelteComponentTag,
  propsResult: PropsResult,
  singleRoot: boolean,
  context: TransformContext<SvelteElement>,
  isDynamicComponent: boolean
) {
  const dynamicComponent = isDynamicComponent ? resolveDynamicComponent(node) : undefined

  let { name: tag } = node
  let asset = true

  if (dynamicComponent) {
    // rename tag name to vapor component tag
    if (tag === 'svelte:component') {
      tag = 'component'
    }
  } else {
    const fromSetup = resolveSetupReference(tag, context)
    if (fromSetup) {
      tag = fromSetup
      asset = false
    }

    const dotIndex = tag.indexOf('.')
    if (dotIndex > 0) {
      const ns = resolveSetupReference(tag.slice(0, dotIndex), context)
      if (ns) {
        tag = ns + tag.slice(dotIndex)
        asset = false
      }
    }

    if (asset) {
      context.component.add(tag)
    }
  }

  context.dynamic.flags |= DynamicFlag.NON_TEMPLATE | DynamicFlag.INSERT
  // TODO:
  // const root = context.root === context.parent && (context.parent.node.children || []).length === 1

  context.registerOperation({
    type: IRNodeTypes.CREATE_COMPONENT_NODE,
    id: context.reference(),
    tag,
    props: propsResult[0] ? propsResult[1] : [propsResult[1]],
    asset,
    root: singleRoot,
    slots: [...context.slots],
    once: context.inVOnce,
    dynamic: dynamicComponent
  })
  context.slots = []
}

function resolveDynamicComponent(node: SvelteComponentTag) {
  const expression = node.expression
  if (!expression) {
    return
  }
  return convertVaporDirectiveComponentExpression(node).exp
}

function resolveSetupReference(name: string, context: TransformContext): string | undefined {
  const bindings = context.options.bindingMetadata
  if (!bindings || bindings.__isScriptSetup === false) {
    return
  }

  const camelName = camelize(name)
  const PascalName = capitalize(camelName)
  return bindings[name]
    ? name
    : bindings[camelName]
      ? camelName
      : bindings[PascalName]
        ? PascalName
        : undefined
}

function transformNativeElement(
  node: SvelteElement,
  propsResult: PropsResult,
  singleRoot: boolean,
  context: TransformContext<SvelteElement>
) {
  const { name: tag } = node
  // TODO: remove
  // const { scopeId } = context.options

  let template = ''
  template += `<${tag}`
  // TODO: remove
  // if (scopeId) {
  //   template += ` ${scopeId}`
  // }

  let staticProps = false
  const dynamicProps: string[] = []
  if (propsResult[0] /* dynamic props */) {
    const [, dynamicArgs, expressions] = propsResult
    context.registerEffect(expressions, {
      type: IRNodeTypes.SET_DYNAMIC_PROPS,
      element: context.reference(),
      props: dynamicArgs,
      root: singleRoot
    })
  } else {
    for (const prop of propsResult[1]) {
      const { key, values } = prop
      if (key.isStatic && values.length === 1 && values[0].isStatic) {
        staticProps = true
        template += ` ${key.content}`
        if (values[0].content) {
          template += `="${values[0].content}"`
        }
      } else {
        dynamicProps.push(key.content)
        context.registerEffect(values, {
          type: IRNodeTypes.SET_PROP,
          element: context.reference(),
          prop,
          root: singleRoot
        })
      }
    }
  }

  if (singleRoot) {
    context.registerOperation({
      type: IRNodeTypes.SET_INHERIT_ATTRS,
      staticProps: staticProps,
      dynamicProps: propsResult[0] ? true : dynamicProps
    })
  }

  template += `>` + context.childrenTemplate.join('')
  // TODO remove unnecessary close tag, e.g. if it's the last element of the template
  if (!isVoidTag(tag)) {
    template += `</${tag}>`
  }

  if (
    context.parent &&
    context.parent.node.type === 'Element' &&
    !isValidHTMLNesting(context.parent.node.name as string, tag)
  ) {
    context.reference()
    context.dynamic.template = context.pushTemplate(template)
    context.dynamic.flags |= DynamicFlag.INSERT | DynamicFlag.NON_TEMPLATE
  } else {
    context.template += template
  }
}

export type PropsResult =
  | [dynamic: true, props: IRProps[], expressions: SimpleExpressionNode[]]
  | [dynamic: false, props: IRPropsStatic]

export function buildProps(
  node: SvelteElement,
  context: TransformContext<SvelteElement>,
  isComponent: boolean,
  isDynamicComponent?: boolean
): PropsResult {
  // context.options.scopedStyleApplyer?.(node, context)

  // convert from svelte props to vapor props
  const props = convertProps(node)
  if (props.length === 0) {
    return [false, []]
  }

  const dynamicArgs: IRProps[] = []
  const dynamicExpr: SimpleExpressionNode[] = []
  let results: DirectiveTransformResult[] = []

  function pushMergeArg() {
    if (results.length > 0) {
      dynamicArgs.push(dedupeProperties(results))
      results = []
    }
  }

  for (const prop of props) {
    if (prop.type === NodeTypes.DIRECTIVE && !prop.arg) {
      if (prop.name === 'bind') {
        // v-bind="obj"
        if (prop.exp) {
          dynamicExpr.push(prop.exp)
          pushMergeArg()
          dynamicArgs.push({
            kind: IRDynamicPropsKind.EXPRESSION,
            value: prop.exp
          })
        } else {
          context.options.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, prop.loc))
        }
        continue
      } else if (prop.name === 'on') {
        // v-on="obj"
        if (prop.exp) {
          if (isComponent) {
            dynamicExpr.push(prop.exp)
            pushMergeArg()
            dynamicArgs.push({
              kind: IRDynamicPropsKind.EXPRESSION,
              value: prop.exp,
              handler: true
            })
          } else {
            context.registerEffect(
              [prop.exp],

              {
                type: IRNodeTypes.SET_DYNAMIC_EVENTS,
                element: context.reference(),
                event: prop.exp
              }
            )
          }
        } else {
          context.options.onError(createCompilerError(ErrorCodes.X_V_ON_NO_EXPRESSION, prop.loc))
        }
        continue
      }
    }

    // TODO:
    // do we need to handle for svelte dynamic component ?
    // svelte dynamic component is not supported statically...

    // exclude `this` prop for <component>
    if (
      (isDynamicComponent && prop.type === NodeTypes.ATTRIBUTE && prop.name === 'this') ||
      (prop.type === NodeTypes.DIRECTIVE && prop.name === 'bind' && isStaticArgOf(prop.arg, 'this'))
    ) {
      continue
    }

    const result = transformProp(prop, node, context)

    if (result) {
      dynamicExpr.push(result.key, result.value)
      if (isComponent && !result.key.isStatic) {
        // v-bind:[name]="value" or v-on:[name]="value"
        pushMergeArg()
        dynamicArgs.push(
          extend(resolveDirectiveResult(result), {
            kind: IRDynamicPropsKind.ATTRIBUTE
          }) as IRPropsDynamicAttribute
        )
      } else {
        // other static props
        results.push(result)
      }
    }
  }

  // has dynamic key or v-bind="{}"
  if (dynamicArgs.length > 0 || results.some(({ key }) => !key.isStatic)) {
    // take rest of props as dynamic props
    pushMergeArg()
    return [true, dynamicArgs, dynamicExpr]
  }

  const irProps = dedupeProperties(results)
  return [false, irProps]
}

function transformProp(
  prop: VaporDirectiveNode | AttributeNode,
  node: SvelteElement,
  context: TransformContext<SvelteElement>
): DirectiveTransformResult | void {
  const { name } = prop

  if (prop.type === NodeTypes.ATTRIBUTE) {
    if (isReservedProp(name)) {
      return
    }
    return {
      key: createSimpleExpression(prop.name, true, prop.nameLoc),
      value: prop.value
        ? createSimpleExpression(prop.value.content, true, prop.value.loc)
        : EMPTY_EXPRESSION
    }
  }

  const directiveTransform = context.options.directiveTransforms[name]
  if (directiveTransform) {
    return directiveTransform(prop, node, context)
  }

  // TODO: should be handled for svelte built-in directives
  // if (!isBuiltInDirective(name)) {
  //   context.directive.add(name)
  //   context.registerOperation({
  //     type: IRNodeTypes.WITH_DIRECTIVE,
  //     element: context.reference(),
  //     dir: prop,
  //     name,
  //   })
  // }
}

// Dedupe props in an object literal.
// Literal duplicated attributes would have been warned during the parse phase,
// however, it's possible to encounter duplicated `onXXX` handlers with different
// modifiers. We also need to merge static and dynamic class / style attributes.
function dedupeProperties(results: DirectiveTransformResult[]): IRProp[] {
  const knownProps: Map<string, IRProp> = new Map()
  const deduped: IRProp[] = []

  for (const result of results) {
    const prop = resolveDirectiveResult(result)
    // dynamic keys are always allowed
    if (!prop.key.isStatic) {
      deduped.push(prop)
      continue
    }
    const name = prop.key.content
    const existing = knownProps.get(name)
    if (existing) {
      if (name === 'style' || name === 'class') {
        mergePropValues(existing, prop)
      }
      // unexpected duplicate, should have emitted error during parse
    } else {
      knownProps.set(name, prop)
      deduped.push(prop)
    }
  }
  return deduped
}

function resolveDirectiveResult(prop: DirectiveTransformResult): IRProp {
  return extend({}, prop, {
    value: undefined,
    values: [prop.value]
  })
}

function mergePropValues(existing: IRProp, incoming: IRProp) {
  const newValues = incoming.values
  existing.values.push(...newValues)
}
