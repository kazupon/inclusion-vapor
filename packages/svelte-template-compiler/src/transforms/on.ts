// SPDX-License-Identifier: MIT
// Modifier: kazuya kawaguchi (a.k.a. kazupon)
// Forked from `@vue/compiler-vapor`
// Author: Evan you (https://github.com/yyx990803) and Vapor team (https://github.com/orgs/vuejs/teams/vapor)
// Repository url: https://github.com/vuejs/core-vapor
// Code url: https://github.com/vuejs/core-vapor/blob/6608bb31973d35973428cae4fbd62026db068365/packages/compiler-vapor/src/transforms/vOn.ts

import { ErrorCodes, createCompilerError, resolveModifiers } from '@vue-vapor/compiler-dom'
import { extend, makeMap } from '@vue-vapor/shared'
import { IRNodeTypes, isSvelteElement } from '../ir/index.ts'
import { EMPTY_EXPRESSION, resolveExpression } from './utils.ts'

import type { KeyOverride, SetEventIRNode } from '../ir/index.ts'
import type { DirectiveTransform } from './types.ts'

const delegatedEvents = /*#__PURE__*/ makeMap(
  'beforeinput,click,dblclick,contextmenu,focusin,focusout,input,keydown,' +
    'keyup,mousedown,mousemove,mouseout,mouseover,mouseup,pointerdown,' +
    'pointermove,pointerout,pointerover,pointerup,touchend,touchmove,' +
    'touchstart'
)

export const transformOn: DirectiveTransform = (dir, node, context) => {
  const { exp, loc, modifiers } = dir
  const isComponent = isSvelteElement(node) && node.type === 'InlineComponent'
  // TODO:
  // const isSlotOutlet = node.tag === 'slot'
  const isSlotOutlet = false

  if (!exp && modifiers.length === 0) {
    context.options.onError(createCompilerError(ErrorCodes.X_V_ON_NO_EXPRESSION, loc))
  }
  let arg = resolveExpression(dir.arg!)

  const { keyModifiers, nonKeyModifiers, eventOptionModifiers } = resolveModifiers(
    arg.isStatic ? `on${arg.content}` : arg,
    modifiers,
    null, // eslint-disable-line unicorn/no-null
    loc
  )

  let keyOverride: KeyOverride | undefined
  const isStaticClick = arg.isStatic && arg.content.toLowerCase() === 'click'
  const delegate = arg.isStatic && eventOptionModifiers.length === 0 && delegatedEvents(arg.content)

  // normalize click.right and click.middle since they don't actually fire
  if (nonKeyModifiers.includes('middle')) {
    if (keyOverride) {
      // TODO error here
    }
    if (isStaticClick) {
      arg = extend({}, arg, { content: 'mouseup' })
    } else if (!arg.isStatic) {
      keyOverride = ['click', 'mouseup']
    }
  }
  if (nonKeyModifiers.includes('right')) {
    if (isStaticClick) {
      arg = extend({}, arg, { content: 'contextmenu' })
    } else if (!arg.isStatic) {
      keyOverride = ['click', 'contextmenu']
    }
  }

  if (isComponent || isSlotOutlet) {
    const handler = exp || EMPTY_EXPRESSION
    return {
      key: arg,
      value: handler,
      handler: true
    }
  }

  const operation: SetEventIRNode = {
    type: IRNodeTypes.SET_EVENT,
    element: context.reference(),
    key: arg,
    value: exp,
    modifiers: {
      keys: keyModifiers,
      nonKeys: nonKeyModifiers,
      options: eventOptionModifiers
    },
    keyOverride,
    delegate,
    effect: !arg.isStatic
  }

  context.registerEffect([arg], operation)
}
