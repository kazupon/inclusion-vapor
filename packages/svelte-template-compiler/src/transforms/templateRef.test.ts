import { NodeTypes } from '@vue-vapor/compiler-dom'
import { compile as vaporCompile } from '@vue-vapor/compiler-vapor'
import { expect, test } from 'vitest'
import { IRNodeTypes } from '../ir/index.ts'
import { makeCompile } from './_utils.ts'
import { transformVBind } from './bind.ts'
import { transformChildren } from './children.ts'
import { transformComment } from './comment.ts'
import { transformElement } from './element.ts'
import { transformVModel } from './model.ts'
import { transformVOn } from './on.ts'
import { transformTemplateRef } from './templateRef.ts'
import { transformText } from './text.ts'

const compileWithTransformRef = makeCompile({
  prefixIdentifiers: false,
  nodeTransforms: [
    transformTemplateRef,
    transformText,
    transformElement,
    transformComment,
    transformChildren
  ],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
    model: transformVModel
  }
})

test('native element', () => {
  const source1 = '<canvas bind:this={el} />'
  const source2 = '<canvas ref="el" />'
  const { code, ir } = compileWithTransformRef(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')

  expect(code).contains(`_setRef(n0, "el")`)

  expect(ir.template).toEqual(['<canvas></canvas>'])
  expect(ir.block.operation).toMatchObject([
    {
      type: IRNodeTypes.SET_INHERIT_ATTRS,
      staticProps: false,
      dynamicProps: []
    },
    {
      type: IRNodeTypes.SET_TEMPLATE_REF,
      element: 0,
      effect: false,
      refFor: false,
      value: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'el',
        isStatic: true,
        loc: {
          // TODO: we need to align svelte AST source location
          source: 'el'
        }
      }
    }
  ])
})

test('component', () => {
  const source1 = '<MyComp bind:this={instance} />'
  const source2 = '<MyComp ref="instance" />'

  const { code, ir: _ } = compileWithTransformRef(source1)
  const expectedResult = vaporCompile(source2)

  expect(code).toMatchSnapshot('svelte')
  expect(expectedResult.code).toMatchSnapshot('vue')
})
