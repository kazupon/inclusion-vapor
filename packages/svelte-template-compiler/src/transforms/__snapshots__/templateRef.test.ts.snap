// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`native element > expected 1`] = `
"import { setRef as _setRef, template as _template } from 'vue/vapor';
const t0 = _template("<canvas></canvas>")

export function render(_ctx) {
  const n0 = t0()
  _setRef(n0, "el")
  return n0
}"
`;

exports[`native element > received 1`] = `
"import { setRef as _setRef, template as _template } from 'vue/vapor';
const t0 = _template("<canvas></canvas>")

export function render(_ctx) {
  const n0 = t0()
  _setRef(n0, "el")
  return n0
}"
`;
